### PR TITLE
WIP: add helm test step to gh-actions

### PR DESCRIPTION
**Pull request by knelasevero**
Also, remove local helm testing in favor of ct in PRs.

Example successful run can be checked here (from my fork to my fork): https://github.com/knelasevero/descheduler/pull/2/checks

Example fail: https://github.com/knelasevero/descheduler/runs/7564334471?check_suite_focus=true

We already get events, describe, and logs of the deployed chart pod in the CI directly. 

I am already removing the scripts and references to previous helm testing approach. 

So /hold this one until I normalize and merge https://github.com/kubernetes/test-infra/pull/26630

### CODE REVIEW

**Comment by knelasevero** [comment]
@cpanato you mentioned some improvements that could be done to this. If you prefer we can review in this PR, or I can merge this and wait for a follow up PR from you. Your call :heart:

File: .github/workflows/helm.yaml [hunk 1]
```diff
@@ -39,6 +43,28 @@ jobs:
...
+          nodes:
+          - role: control-plane
+          - role: worker
+          - role: worker
+          EOF
+          kind create cluster --name kind --config kind-config.yaml --wait 2m
+
+      # helm-extra-set-args only available after ct 3.6.0
+      - name: Run chart-testing (install)
+        run: ct install --config=.github/ci/ct.yaml --helm-extra-set-args="--set=kind=Deployment --set=podSecurityPolicy.create=false"
```
Comment (ingvagabund) [inline .github/workflows/helm.yaml:70]: Would be great to extend the list of steps with running the e2e

File: docs/contributor-guide.md [hunk 2]
```diff
@@ -39,19 +39,6 @@ make test-unit
...
-Run the helm test for a particular descheduler release by setting below variables,
-```
-HELM_IMAGE_REPO="descheduler"
-HELM_IMAGE_TAG="helm-test"
-HELM_CHART_LOCATION="./charts/descheduler"
-```
-The helm tests runs as part of descheduler CI. But, to run it manually from the descheduler root,
-
-```
-make test-helm
```
Comment (ingvagabund) [inline docs/contributor-guide.md:52]: Do we have an alternative to running the helm test locally? Can `.github/workflows/helm.yaml` by run the same way through `make test-helm`?

**Review by ingvagabund — COMMENTED** [review]
This looks really great. I like how easy it is to read the workflow. Also https://github.com/kubernetes-sigs/descheduler/runs/7579750658?check_suite_focus=true looks great with all the events, pod spec and container logs.

`Helm / lint-and-test (pull_request)` test is currently not required. We need to make sure it is.

File: docs/contributor-guide.md (hunk 2 above)
Comment (knelasevero) [reply docs/contributor-guide.md:52]: I'm figuring out how to run ct locally without a lot of dependencies, will add that here :)

File: .github/workflows/helm.yaml (hunk 1 above)
Comment (knelasevero) [reply .github/workflows/helm.yaml:70]: Makes sense!

Comment (cpanato) [inline .github/workflows/helm.yaml:70]: you dont need to set the `"--set=kind=Deployment --set=podSecurityPolicy.create=false"` you can create a ci directory and add the values there, see for example: https://github.com/sigstore/helm-charts/tree/main/charts/fulcio/ci

**Review by cpanato — COMMENTED** [review]
https://github.com/kubernetes-sigs/descheduler/pull/895/files#diff-3be36a9541f01c2e8a31c46ccbbb49265f35f3dc8bb57d01160d4f925961a87cR25-R28

also can be bumped to use helm `3.9.2`

File: .github/workflows/helm.yaml (hunk 1 above)
Comment (knelasevero) [reply .github/workflows/helm.yaml:70]: Nice, thanks!

**Comment by knelasevero** [comment]
> https://github.com/kubernetes-sigs/descheduler/pull/895/files#diff-3be36a9541f01c2e8a31c46ccbbb49265f35f3dc8bb57d01160d4f925961a87cR25-R28
> 
> also can be bumped to use helm `3.9.2`

Will do!

File: .github/workflows/helm.yaml [hunk 3]
```diff
@@ -39,6 +43,35 @@ jobs:
           if [[ -n "$changed" ]]; then
             echo "::set-output name=changed::true"
           fi
+      
+      # Force linting/test in debug runs with a change (this action only runs on chart and helm.yaml workflow diffs anyways)
+      - name: Change Kind to deployment
+        run: "sed -i \"s|kind: [^ ]*|kind: Deployment|g\" charts/descheduler/values.yaml"
```
Comment (a7i) [inline .github/workflows/helm.yaml:49]: this seems odd to do, when we can have a specific values file that can do this.

File: .github/workflows/helm.yaml [hunk 4]
```diff
@@ -39,6 +43,35 @@ jobs:
...
+          cat > kind-config.yaml <<EOF
+          # three node (two workers) cluster config
+          kind: Cluster
+          apiVersion: kind.x-k8s.io/v1alpha4
+          nodes:
+          - role: control-plane
+          - role: worker
+          - role: worker
+          EOF
+          kind create cluster --name kind --config kind-config.yaml --wait 2m
```
Comment (a7i) [inline .github/workflows/helm.yaml:66]: Can this not reference the [kind_config](https://github.com/kubernetes-sigs/descheduler/blob/master/hack/kind_config.yaml) used by e2e?

Comment (a7i) [reply .github/workflows/helm.yaml:66]: for topology spread tests, we label the nodes with unique AZ values

File: .github/workflows/helm.yaml (hunk 3 above)
Comment (knelasevero) [reply .github/workflows/helm.yaml:49]: This was only to force ct lint and ct install to show changes, it is unrelated to this value change itself. I want to remove this when the work is finished.

Comment (knelasevero) [reply .github/workflows/helm.yaml:49]: Initially I was planning on leaving this, but now I see it is too ugly/confusing :sweat_smile:

File: .github/workflows/helm.yaml (hunk 4 above)
Comment (knelasevero) [reply .github/workflows/helm.yaml]: Great! using it now.

File: hack/verify-chart.sh [hunk 5]
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+
+docker run -it --rm --network host --workdir=/data --volume ~/.kube/config:/root/.kube/config:ro --volume $(pwd):/data quay.io/helmpack/chart-testing:v3.7.0 /bin/bash -c "git config --global --add safe.directory /data; ct install --config=.github/ci/ct.yaml --helm-extra-set-args=\"--set=kind=Deployment --set=podSecurityPolicy.create=false\""
```
Comment (knelasevero) [inline hack/verify-chart.sh:3]: For some reason when calling from make the helm-extra-set-args from .github/ci/ct.yaml are being ignored... had to set them manually here as arguments of ct.

File: docs/contributor-guide.md (hunk 2 above)
Comment (knelasevero) [reply docs/contributor-guide.md:52]: Threw some things on Makefile, docs and ./hack/verify-helm.sh

**Comment by knelasevero** [comment]
/hold after reviews are done I will remove https://github.com/kubernetes-sigs/descheduler/pull/895#discussion_r934597809

and also need to wait for https://github.com/kubernetes/test-infra/pull/26630

**Comment by ingvagabund** [comment]
> and also need to wait for https://github.com/kubernetes/test-infra/pull/26630

I'd prefer first to introduce a new test, merge this PR, merge https://github.com/kubernetes/test-infra/pull/26630 and then change the `make test-helm` to run locally.

**Comment by knelasevero** [comment]
> > and also need to wait for [kubernetes/test-infra#26630](https://github.com/kubernetes/test-infra/pull/26630)
> 
> I'd prefer first to introduce a new test, merge this PR, merge [kubernetes/test-infra#26630](https://github.com/kubernetes/test-infra/pull/26630) and then change the `make test-helm` to run locally.

New test? 🤔 You mean this PR itself as the replacement, or something else?

> Merge this PR

If we merge this PR first, all new PRs will fail because of inexistent scripts being called on test-infra steps.

> make test-helm

I've introduced make ct-helm in this PR, and it runs locally against kind

> In overall, this is quite awesome improvement :)

🙌🙌

**Comment by ingvagabund** [comment]
I would like to see the changes in the following order:
1. updating this PR so only a new `Helm / lint-and-test (pull_request)` test is introduced while still keeping `pull-descheduler-helm-test` and `make test-helm` unchanged (merged through https://github.com/kubernetes-sigs/descheduler/pull/895)
2. making sure the new `Helm / lint-and-test (pull_request)` test is required and a PR does not merge without the new test going green (this is paramount and needs to be verified)
3. removing `pull-descheduler-helm-test` test (https://github.com/kubernetes/test-infra/pull/26630 gets merged)
4. updating the `make test-helm` to run the new `Helm / lint-and-test (pull_request)` test locally (new PR)

**Comment by knelasevero** [comment]
> I would like to see the changes in the following order:
> 
> 1. updating this PR so only a new `Helm / lint-and-test (pull_request)` test is introduced while still keeping `pull-descheduler-helm-test` and `make test-helm` unchanged (merged through [add helm test step to gh-actions #895](https://github.com/kubernetes-sigs/descheduler/pull/895))
> 2. making sure the new `Helm / lint-and-test (pull_request)` test is required and a PR does not merge without the new test going green (this is paramount and needs to be verified)
> 3. removing `pull-descheduler-helm-test` test ([descheduler: remove helm tests from prow kubernetes/test-infra#26630](https://github.com/kubernetes/test-infra/pull/26630) gets merged)
> 4. updating the `make test-helm` to run the new `Helm / lint-and-test (pull_request)` test locally (new PR)

You are right, that makes sense. I will convert this PR in a draft/wip to keep it as reference, and open new PRs step by step then. Then we can actually just test if prow ignores or considers gh-actions workflows before merging right after the first merge :)

**Pull request closed without merging**